File: target_py3/cum/cum.py
"""
CUM Python 3 runtime — mirrors `target_js/cum/cum.mjs` / `target_cpp/cum/cum.hpp`.

Packed primitives and PER-style packing rules match the JS implementation.
"""

from __future__ import annotations

import struct
from typing import SupportsIndex, Union

BufferLike = Union[bytes, bytearray, memoryview]


class CodecError(Exception):
    """Invalid encode/decode inputs or malformed bitstream."""

    pass


def octets_for_cum_capacity(capacity: SupportsIndex) -> int:
    n = int(capacity)
    if n < 1:
        raise CodecError("invalid capacity {!r}".format(capacity))
    if n <= 256:
        return 1
    if n <= 65536:
        return 2
    if n <= 4294967296:
        return 4
    return 8


def octets_for_choice_arity(alternatives: SupportsIndex) -> int:
    n = int(alternatives)
    if n < 2:
        raise CodecError("choice must have ≥2 alternatives")
    if n < 256:
        return 1
    if n < 65536:
        return 2
    if n < 4294967296:
        return 4
    return 8


def set_optional(mask: bytearray, idx: int) -> None:
    byte = idx >> 3
    bit = idx & 7
    mask[byte] |= 0x80 >> bit


def check_optional(mask: Union[bytes, bytearray, memoryview], idx: int) -> bool:
    byte = idx >> 3
    bit = idx & 7
    return (mask[byte] & (0x80 >> bit)) != 0


def write_integral_le(buf: Union[bytes, bytearray, memoryview], dst_off: int, value: int, nbytes: int) -> None:
    v = int(value) % (1 << (nbytes * 8))
    for b in range(nbytes):
        buf[dst_off + b] = (v >> (8 * b)) & 0xFF


def read_integral_le(buf: BufferLike, src_off: int, nbytes: int) -> int:
    v = 0
    mv = buf if isinstance(buf, memoryview) else memoryview(buf)
    for b in range(nbytes):
        v |= int(mv[src_off + b]) << (8 * b)
    return v


class PerCodecCtx:
    """Packed codec cursor; corresponds to `cum::per_codec_ctx`."""

    __slots__ = ("buf", "off")

    buf: BufferLike
    off: int

    def __init__(self, backing: BufferLike, offset: int = 0):
        self.buf = backing
        self.off = int(offset)

    def remaining(self) -> int:
        return len(self.buf) - self.off

    def _bump(self, n: int) -> None:
        if self.remaining() < n:
            raise CodecError("codec attempted past end of buffer")
        self.off += n

    def write_bytes(self, src: BufferLike, length: int | None = None) -> None:
        if isinstance(src, memoryview):
            sl = len(src)
        else:
            sl = len(src)
        ln = sl if length is None else length
        if ln > sl:
            raise CodecError("write slice too long")
        if ln > self.remaining():
            raise CodecError("encode buffer full")
        mo = memoryview(src)
        self.buf[self.off : self.off + ln] = mo[:ln]
        self.off += ln

    def read_bytes(self, ln: int) -> bytes:
        if ln > self.remaining():
            raise CodecError("decode overrun")
        mv = memoryview(self.buf)
        chunk = mv[self.off : self.off + ln].tobytes()
        self.off += ln
        return chunk

    def write_i32le(self, v: int) -> None:
        if self.remaining() < 4:
            raise CodecError("encode buffer full")
        if isinstance(self.buf, bytearray):
            struct.pack_into("<i", self.buf, self.off, int(v))
        else:
            raise CodecError("encode requires a writable bytearray backing")
        self.off += 4

    def read_i32le(self) -> int:
        if self.remaining() < 4:
            raise CodecError("decode overrun")
        mv = memoryview(self.buf)
        val = struct.unpack_from("<i", mv, self.off)[0]
        self.off += 4
        return int(val)

    def write_u8(self, byte: int) -> None:
        if self.remaining() < 1:
            raise CodecError("encode buffer full")
        if not isinstance(self.buf, bytearray):
            raise CodecError("encode requires a writable bytearray backing")
        self.buf[self.off] = int(byte) & 0xFF
        self.off += 1

    def read_u8(self) -> int:
        if self.remaining() < 1:
            raise CodecError("decode overrun")
        b = memoryview(self.buf)[self.off]
        self.off += 1
        return int(b)

    def write_count(self, max_cardinality: int, count: int) -> None:
        nb = octets_for_cum_capacity(max_cardinality)
        cap = int(max_cardinality)
        if count < 0 or count > cap:
            raise CodecError(
                "collection count {} out of range (max {})".format(count, cap)
            )
        if self.remaining() < nb:
            raise CodecError("encode buffer full")
        if not isinstance(self.buf, bytearray):
            raise CodecError("encode requires a writable bytearray backing")
        write_integral_le(self.buf, self.off, int(count), nb)
        self.off += nb

    def read_count(self, max_cardinality: int) -> int:
        nb = octets_for_cum_capacity(max_cardinality)
        cap = int(max_cardinality)
        if self.remaining() < nb:
            raise CodecError("decode overrun")
        count = read_integral_le(self.buf, self.off, nb)
        self.off += nb
        if count > cap or count < 0:
            raise CodecError("decoded count {} out of range (max {})".format(count, cap))
        return count

    def encode_c_string_latin1(self, string: str) -> None:
        n = len(string) + 1
        if n > self.remaining():
            raise CodecError("encode buffer full")
        if not isinstance(self.buf, bytearray):
            raise CodecError("encode requires a writable bytearray backing")
        for ch in string:
            cp = ord(ch)
            if cp > 255:
                raise CodecError("non Latin-1 string not supported for CUM string")
            self.buf[self.off] = cp
            self.off += 1
        self.buf[self.off] = 0
        self.off += 1

    def decode_c_string_latin1(self) -> str:
        mv = memoryview(self.buf)
        start = self.off
        end = start
        max_len = len(mv)
        while end < max_len and mv[end] != 0:
            end += 1
        if end >= max_len:
            raise CodecError("unterminated C string")
        raw = mv[start:end].tobytes()
        self.off = end + 1
        return raw.decode("latin1")

    def write_choice_index(self, num_alternatives: int, index: int) -> None:
        nb = octets_for_choice_arity(num_alternatives)
        if index < 0 or index >= num_alternatives:
            raise CodecError(
                "choice index {!r} invalid ({!r} alts)".format(index, num_alternatives)
            )
        if self.remaining() < nb:
            raise CodecError("encode buffer full")
        if not isinstance(self.buf, bytearray):
            raise CodecError("encode requires a writable bytearray backing")
        write_integral_le(self.buf, self.off, index, nb)
        self.off += nb

    def read_choice_index(self, num_alternatives: int) -> int:
        nb = octets_for_choice_arity(num_alternatives)
        if self.remaining() < nb:
            raise CodecError("decode overrun")
        index = read_integral_le(self.buf, self.off, nb)
        self.off += nb
        if index >= num_alternatives or index < 0:
            raise CodecError("bad choice index {}".format(index))
        return index

File: target_js/cum/cum.mjs
/**
 * CUM JavaScript runtime — mirrors `target_cpp/cum/cum.hpp`.
 *
 * Packed (byte-aligned PER-style) primitives live here; other coding rules can be
 * added to this same module alongside, like the single-header C++ layout.
 */

export class CodecError extends Error {
  constructor(message) {
    super(message);
    this.name = "CodecError";
  }
}

/**
 * Prefix width for lengths/counts capped at `capacity` (dynamic_array / buffer cap).
 * Mirrors cum.hpp `DetermineUnsignedType<N>`: N<=256 -> 1 byte, etc.
 */
export function octetsForCumCapacity(capacity) {
  const N = Number(capacity);
  if (!(N >= 1) || Number.isNaN(N)) throw new CodecError(`invalid capacity ${capacity}`);
  if (N <= 256) return 1;
  if (N <= 65536) return 2;
  if (N <= 4294967296) return 4;
  return 8;
}

/** Discriminator octets for variant index; mirrors CppGenerator.determineUnsignedSize(len(cs)). */
export function octetsForChoiceArity(alternatives) {
  const n = Number(alternatives);
  if (!(n >= 2)) throw new CodecError("choice must have ≥2 alternatives");
  if (n < 256) return 1;
  if (n < 65536) return 2;
  if (n < 4294967296) return 4;
  return 8;
}

export function setOptional(mask, idx) {
  const byte = idx >> 3;
  const bit = idx & 7;
  mask[byte] |= 0x80 >> bit;
}

export function checkOptional(mask, idx) {
  const byte = idx >> 3;
  const bit = idx & 7;
  return (mask[byte] & (0x80 >> bit)) !== 0;
}

export function writeIntegralLE(dst, dstOff, value, nbytes) {
  let v = BigInt(value >>> 0);
  if (nbytes < 8) {
    const max = BigInt(1) << BigInt(nbytes * 8);
    v = v & (max - BigInt(1));
  }
  for (let b = 0; b < nbytes; b++) {
    dst[dstOff + b] = Number(v & BigInt(0xff));
    v >>= BigInt(8);
  }
}

export function readIntegralLE(src, srcOff, nbytes) {
  let v = BigInt(0);
  for (let b = 0; b < nbytes; b++) {
    v |= BigInt(src[srcOff + b]) << BigInt(8 * b);
  }
  return Number(v);
}

/** Packed codec cursor; corresponds to cum::per_codec_ctx + encode_per overloads. */
export class PerCodecCtx {
  /**
   * @param {Uint8Array} backing
   * @param {number} offset
   * @param {"encode" | "decode"} mode
   */
  constructor(backing, offset = 0, mode = "encode") {
    this.buf = backing;
    this.off = offset;
    this.mode = mode;
  }

  remaining() {
    return this.buf.byteLength - this.off;
  }

  bump(n) {
    if (this.remaining() < n) {
      throw new CodecError(`${this.mode} attempted past end of buffer`);
    }
    this.off += n;
  }

  writeBytes(src, len = src.byteLength) {
    if (len > src.byteLength) throw new CodecError("write slice too long");
    if (len > this.remaining()) throw new CodecError("encode buffer full");
    this.buf.set(src.subarray(0, len), this.off);
    this.off += len;
  }

  readBytes(len) {
    if (len > this.remaining()) throw new CodecError("decode overrun");
    const out = this.buf.slice(this.off, this.off + len);
    this.off += len;
    return out;
  }

  writeI32LE(v) {
    if (this.remaining() < 4) throw new CodecError("encode buffer full");
    writeIntegralLE(this.buf, this.off, v, 4);
    this.off += 4;
  }

  readI32LE() {
    if (this.remaining() < 4) throw new CodecError("decode overrun");
    const v = readIntegralLE(this.buf, this.off, 4);
    this.off += 4;
    /** @type {number} */
    return v | 0;
  }

  writeU8(byte) {
    if (this.remaining() < 1) throw new CodecError("encode buffer full");
    this.buf[this.off++] = byte & 0xff;
  }

  readU8() {
    if (this.remaining() < 1) throw new CodecError("decode overrun");
    return this.buf[this.off++];
  }

  writeCount(maxCardinality, count) {
    const nb = octetsForCumCapacity(maxCardinality);
    const cap = Number(maxCardinality);
    if (!(count >= 0) || count > cap) {
      throw new CodecError(`collection count ${count} out of range (max ${cap})`);
    }
    if (this.remaining() < nb) throw new CodecError("encode buffer full");
    writeIntegralLE(this.buf, this.off, count, nb);
    this.off += nb;
  }

  readCount(maxCardinality) {
    const nb = octetsForCumCapacity(maxCardinality);
    const cap = Number(maxCardinality);
    if (this.remaining() < nb) throw new CodecError("decode overrun");
    const count = readIntegralLE(this.buf, this.off, nb);
    this.off += nb;
    if (count > cap || count < 0) {
      throw new CodecError(`decoded count ${count} out of range (max ${cap})`);
    }
    return count;
  }

  encodeCStringLatin1(str) {
    const n = str.length + 1;
    if (n > this.remaining()) throw new CodecError("encode buffer full");
    for (let i = 0; i < str.length; i++) {
      const cp = str.charCodeAt(i);
      if (cp > 255) throw new CodecError("non Latin-1 string not supported for CUM string");
      this.buf[this.off++] = cp;
    }
    this.buf[this.off++] = 0;
  }

  decodeCStringLatin1() {
    let end = this.off;
    while (end < this.buf.byteLength && this.buf[end] !== 0) end++;
    if (end >= this.buf.byteLength) throw new CodecError("unterminated C string");
    const slice = this.buf.subarray(this.off, end);
    this.off = end + 1;
    let s = "";
    for (let i = 0; i < slice.byteLength; i++) s += String.fromCharCode(slice[i]);
    return s;
  }

  writeChoiceIndex(numAlternatives, index) {
    const nb = octetsForChoiceArity(numAlternatives);
    if (!(index >= 0) || index >= numAlternatives) {
      throw new CodecError(`choice index ${index} invalid (${numAlternatives} alts)`);
    }
    if (this.remaining() < nb) throw new CodecError("encode buffer full");
    writeIntegralLE(this.buf, this.off, index, nb);
    this.off += nb;
  }

  readChoiceIndex(numAlternatives) {
    const nb = octetsForChoiceArity(numAlternatives);
    if (this.remaining() < nb) throw new CodecError("decode overrun");
    const index = readIntegralLE(this.buf, this.off, nb);
    this.off += nb;
    if (index >= numAlternatives || index < 0) throw new CodecError(`bad choice index ${index}`);
    return index;
  }
}
